docs/machine-config-operator-certificates: Expand significantly (#61262)
openshift/openshift-docs: security/certificate_types_descriptions/machine-config-operator-certificates.adoc

== Purpose

This certificate authority is used to secure connections from nodes to Machine Config Server (MCS) during initial provisioning.

There are two certificates:
. A self-signed CA, the MCS CA
. A derived certificate, the MCS cert

=== Provisioning details

> {product-title} installations using {op-system-first} are installed using Ignition. This process is split into two parts:

Reviewer: ```suggestion
{product-title} installations that use {op-system-first} are installed by using Ignition. This process is split into two parts:
```
From IBM Style Word Usage:
>using _noun_
To avoid ambiguity, replace this gerund with either _by using_ or _that uses_.

> . An Ignition config is created which is references a URL for the full configuration served by the MCS.

Reviewer: ```suggestion
. An Ignition config is created that references a URL for the full configuration that is served by the MCS.
```
Using "that" instead of "which", as I think this sentence contains a restrictive clause. Source: https://www.ibm.com/docs/en/ibm-style?topic=punctuation-commas#commas-with-nonrestrictive-clauses-and-phrases

> . The Ignition config manifests as a `worker.ign` file created by the `openshift-install` command for user-provisioned infrastucture installation methods. For installer-provisioned infrastructure installation methods that use the Machine API Operator, this configuration appears as the `worker-user-data` secret.

Reviewer: ```suggestion
. For user-provisioned infrastructure installation methods, the Ignition config manifests as a `worker.ign` file created by the `openshift-install` command. For installer-provisioned infrastructure installation methods that use the Machine API Operator, this configuration is displayed as the `worker-user-data` secret.
```
I'm not completely sure what "appears" meant in the original text, but I replaced it with "is displayed" per IBM Style, which warns against using "appear" instead of another term, such as "opens" or "is displayed". If "appears" is the best word to use here considering the context, feel free to use it.

include::snippets/mcs-endpoint-limitation.adoc[]

.Additional resources

* xref:../../post_installation_configuration/machine-configuration-tasks.adoc#understanding-the-machine-config-operator[Understanding the Machine Config Operator].

* xref:../../networking/openshift_sdn/about-openshift-sdn.adoc#about-openshift-sdn[About the OpenShift SDN network plugin].

=== Provisioning chain of trust

> The MCS CA is injected into the Ignition configuration under the `security.tls.certificateAuthorities` configuration field. The MCS then provides the complete configuration with a webserver using the MCS cert.

Reviewer: In the second sentence, the phrase "with a webserver using the MCS cert" is unclear. Does it mean that the MCS provides the complete configuration by using the webserver and the MCE cert?

Author: I think it's "using the MCS cert that is hosted on the web server", but i'm not totally sure. @cgwalters can you weigh in here?

> The MCS CA is injected into the Ignition configuration under the `security.tls.certificateAuthorities` configuration field. The MCS then provides the complete configuration with a webserver using the MCS cert.

Reviewer: ```suggestion
The MCS CA is injected into the Ignition configuration under the `security.tls.certificateAuthorities` configuration field. The MCS then provides the complete configuration with a web server using the MCS cert.
```

> The client, Ignition running on a machine in the initramfs, validates that the MCS cert presented by the server has a chain of trust to an authority, the MCS CA, it recognizes. This ensures that the client is accessing the correct server.

Reviewer: I'm really not sure what the first sentence on line 36 means. Also, should "initramfs" be in monospace font? I've seen that term referenced as a file name before.

Author: @lahinson I am trying to describe using the client/server architecture how certificates get validated and used. Is there a particular part that's unclear? It is kind of a mouthful. Is this rewording any more clear:

> The client validates that the MCS cert presented by the server has a chain of trust to an authority, the MCS CA, it recognizes. This ensure that the client is a    ccessing the correct server. The client in this case is Ignition running on a machine in the initramfs.

I think in this situation it should remain as is. I'm using it like it's used here: https://docs.openshift.com/container-platform/4.13/architecture/architecture-rhcos.html#ignition-sequence_architecture-rhcos. in some cases initramfs is used to refer to the specific collection of files for installation, `initramfs` files, and that's when i see the term monospaced...I think these situations are different

Reviewer: @jldohmann Thanks for the explanation! Your revision is better, although I wonder if the end of the first sentence can be clearer. Would something like this work?

>The client validates that the MCS cert that the server presented has a chain of trust to an authority that the MCS cert recognizes, such as the MCS CA. This ensures that the client is accessing the correct server. In this case, the client is Ignition running on a machine in the initramfs.

Reviewer 2: > The client validates that the MCS cert that the server presented has a chain of trust to an authority that the MCS cert recognizes, such as the MCS CA. 

It's the other way around, the MCS CA signs the MCS cert.  I feel my initial wording here was reasonably accurate for what it's worth.

See also the "security.tls.certificateAuthorities" section under https://coreos.github.io/ignition/configuration-v3_4/ which is how this works - the MCS CA is included in that value.

Reviewer 2: I recommend a glance at https://en.wikipedia.org/wiki/Certificate_authority btw

Reviewer: @cgwalters Thanks for the clarification and the resources. In the initial wording that I reviewed, which I've included below, I had a hard time figuring out what the two phrases that are surrounded by commas ("Ignition running on a machine in the initramfs" and "the MCS CA") were modifying. 

>The client, Ignition running on a machine in the initramfs, validates that the MCS cert presented by the server has a chain of trust to an authority, the MCS CA, it recognizes.

Jesse's suggested revision clarified the meaning of the first phrase. However, as I'm sure you could tell based on my last comment, I'm still having trouble understanding how the phrase "the MCS CA" fits in. 

I appreciate your help to get this right. Is this revision more accurate?

>The MCS CA signs the MCS cert. The client uses the MCS cert to authenticate the signature and validates that the MCA cert that the server presents has a chain of trust to a recognized authority.

Author: I've updated the wording a little bit

Reviewer: LTGM! Thanks for your patience in considering all of this feedback.

=== Key material inside a cluster

> The MCS CA appears in the cluster as a config map under the `kube-system` namespace, object `root-ca` and with key `ca.crt`.  The private key is not stored in the cluster and is discarded after the installation completes.

Reviewer: ```suggestion
The MCS CA is displayed in the cluster as a config map under the `kube-system` namespace, `root-ca` object, with `ca.crt` key.  The private key is not stored in the cluster and is discarded after the installation completes.
```
See my earlier comment about "appears". Also, I'm not sure if "under the namespace" is correct or if a different preposition might be clearer (in, on, etc.)

> The MCS cert appears in the cluster as a secret under the `openshift-machine-config-operator` namespace, object `machine-config-server-tls` with keys `tls.crt` and `tls.key`.

Reviewer: ```suggestion
The MCS cert is displayed in the cluster as a secret under the `openshift-machine-config-operator` namespace and `machine-config-server-tls` object with the `tls.crt` and `tls.key` keys.
```
Same comments from line 40 apply here.

== Management

> At this time, modifying either of these certificates directly is not supported.

Reviewer: ```suggestion
At this time, directly modifying either of these certificates is not supported.
```
In most cases, place an adverb directly before the word it modifies.
